src/server/coprocessor/executor/base_scanner.rs

```diff
@@ -0,0 +1,394 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use server::coprocessor::Result;
+use kvproto::coprocessor::KeyRange;
+use storage::{Key, Value, SnapshotStore, Statistics, ScanMode};
+use util::escape;
+
+// Scanner for TableScan and IndexScan
+pub struct BaseScanner<'a> {
+    store: SnapshotStore<'a>,
+    seek_key: Option<Vec<u8>>,
+    scan_mode: ScanMode,
+    upper_bound: Option<Vec<u8>>,
+    statistics: &'a mut Statistics,
+    desc: bool,
+    key_only: bool,
+}
+
+
+impl<'a> BaseScanner<'a> {
+    pub fn new(desc: bool,
+               key_only: bool,
+               store: SnapshotStore<'a>,
+               statistics: &'a mut Statistics)
+               -> BaseScanner<'a> {
+
+        let scan_mode = if desc {
+            ScanMode::Backward
+        } else {
+            ScanMode::Forward
+        };
+        BaseScanner {
+            store: store,
+            seek_key: None,
```

Comment:
1. Since `seek_key`'s share the same logic `prepare_and_get_seek_key` in `index_scan` and `table_scan`, that's way I put it into `scanner` .

2. The name `get_row_from_range` is copied from `tidb`'s `getRowFromRange`. If we put `seek_key` in the `scanner`, I prefer use the same name with `tidb` @coocood